server/schedule/operator/builder.go

```diff
@@ -415,49 +597,73 @@ func (b *Builder) planReplace() stepPlan {
 	var best stepPlan
 	// add voter + remove voter OR add learner + remove learner.
 	for _, i := range b.toAdd.IDs() {
-		add := b.toAdd.Get(i)
+		add := b.toAdd[i]
 		for _, j := range b.toRemove.IDs() {
-			remove := b.toRemove.Get(j)
+			remove := b.toRemove[j]
 			if core.IsLearner(remove) == core.IsLearner(add) {
 				best = b.planReplaceLeaders(best, stepPlan{add: add, remove: remove})
 			}
 		}
 	}
 	// add learner + promote learner + remove voter
 	for _, i := range b.toPromote.IDs() {
-		promote := b.toPromote.Get(i)
+		promote := b.toPromote[i]
 		for _, j := range b.toAdd.IDs() {
-			if add := b.toAdd.Get(j); core.IsLearner(add) {
+			if add := b.toAdd[j]; core.IsLearner(add) {
 				for _, k := range b.toRemove.IDs() {
-					if remove := b.toRemove.Get(k); !core.IsLearner(remove) && j != k {
+					if remove := b.toRemove[k]; !core.IsLearner(remove) && j != k {
 						best = b.planReplaceLeaders(best, stepPlan{promote: promote, add: add, remove: remove})
 					}
 				}
 			}
 		}
 	}
+	// demote voter + remove learner + add voter
+	for _, i := range b.toDemote.IDs() {
+		demote := b.toDemote[i]
+		for _, j := range b.toRemove.IDs() {
+			if remove := b.toRemove[j]; core.IsLearner(remove) {
+				for _, k := range b.toAdd.IDs() {
+					if add := b.toAdd[k]; !core.IsLearner(add) && j != k {
+						best = b.planReplaceLeaders(best, stepPlan{demote: demote, add: add, remove: remove})
+					}
+				}
+			}
+		}
+	}
 	return best
 }
 
 func (b *Builder) planReplaceLeaders(best, next stepPlan) stepPlan {
 	// Brute force all possible leader combinations to find the best plan.
-	for _, leaderAdd := range b.currentPeers.IDs() {
-		if !b.allowLeader(b.currentPeers.Get(leaderAdd)) {
+	for _, leaderBeforeAdd := range b.currentPeers.IDs() {
+		if !b.allowLeader(b.currentPeers[leaderBeforeAdd]) {
 			continue
 		}
-		next.leaderAdd = leaderAdd
-		for _, leaderRemove := range b.currentPeers.IDs() {
-			if b.allowLeader(b.currentPeers.Get(leaderRemove)) && leaderRemove != next.remove.GetStoreId() {
-				next.leaderRemove = leaderRemove
+		next.leaderBeforeAdd = leaderBeforeAdd
+		for _, leaderBeforeRemove := range b.currentPeers.IDs() {
+			if leaderBeforeRemove != next.demote.GetStoreId() &&
+				leaderBeforeRemove != next.remove.GetStoreId() &&
+				b.allowLeader(b.currentPeers[leaderBeforeRemove]) {
+				// leaderBeforeRemove does not select nodes to be demote or removed.
+				next.leaderBeforeRemove = leaderBeforeRemove
 				best = b.comparePlan(best, next)
 			}
 		}
-		if next.promote != nil && b.allowLeader(next.promote) && next.promote.GetStoreId() != next.remove.GetStoreId() {
-			next.leaderRemove = next.promote.GetStoreId()
+		if next.promote != nil &&
+			next.promote.GetStoreId() != next.demote.GetStoreId() &&
+			next.promote.GetStoreId() != next.remove.GetStoreId() &&
+			b.allowLeader(next.promote) {
+			// leaderBeforeRemove does not select nodes to be demote or removed.
+			next.leaderBeforeRemove = next.promote.GetStoreId()
```

Comment:
I' wondering for the case `// add learner + promote learner + remove voter` , will the code panic here since ` next.demote` is nil?

Reply:
`kvproto`'s `GetXXX()` allows `nil` calls.